### ISSUE TITLE
AST access to footnotes and refs

### ISSUE THREAD

**Issue by lread** [asciidoctor/asciidoctorj#958]
I am experimenting with [writing a converter](https://github.com/asciidoctor/asciidoctorj/blob/master/docs/integrator-guide.adoc#publish-everywhere-adapt-asciidoctor-to-your-own-target-format) with AsciidoctorJ (thanks for providing the capability!).

I am using [htm5.rb](https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/converter/html5.rb) as a reference. 

There are currently 2 items that the Ruby code is using that I don't see mirrored in AsciidoctorJ AST classes.

1. footnotes - [used to render the footnotes at the bottom of a document](https://github.com/asciidoctor/asciidoctor/blob/cabde9c25fe040221ad71bb6aaf8754d2e5ec099/lib/asciidoctor/converter/html5.rb#L327).
2. refs - [used, in a way my limited Ruby grokking skills does not fully understand yet, in inline anchors](https://github.com/asciidoctor/asciidoctor/blob/cabde9c25fe040221ad71bb6aaf8754d2e5ec099/lib/asciidoctor/converter/html5.rb#L1115).

Should these be exposed to the AsciiDoctorJ AST? 
Or maybe they are available in some way I have missed?
Are there other similar values that should be exposed?

If exposing these values makes sense, I am happy to do the work. 
To me they look like they should be accessible from [Document](https://github.com/asciidoctor/asciidoctorj/blob/master/asciidoctorj-api/src/main/java/org/asciidoctor/ast/Document.java).
Lemme know!

**Comment by abelsromero** [comment]
I'd say...

> Should these be exposed to the AsciiDoctorJ AST?

Yes

> Or maybe they are available in some way I have missed?

Does not seem so.

> Are there other similar values that should be exposed?

On one side, if we check the mention on `:catalog_assets:` in the documentation: https://asciidoctor.org/docs/user-manual/#ruby-api-options maybe it would make sense to expose the whole `references` property as `Document::getReferences()`. Note that references is a alias for the internal property `catalog`.
But on the other, the core API exposes footnotes and other "reference" properties directly also from Document https://github.com/asciidoctor/asciidoctor/blob/cabde9c25fe040221ad71bb6aaf8754d2e5ec099/lib/asciidoctor/document.rb#L616-L626.

If you ask me, I'd rather have a reference object (ie. `document.getReferences().getFootnotes()`). Having properties grouped helps keeping the API tidier, but I admit can be harder to use if you don't know where to look.


> If exposing these values makes sense, I am happy to do the work.
> To me they look like they should be accessible from [Document](https://github.com/asciidoctor/asciidoctorj/blob/master/asciidoctorj-api/src/main/java/org/asciidoctor/ast/Document.java).
> Lemme know!

You are more than welcome! If you check the code, you'll see the API exposes the featues from the underlying Ruby engine using interfaces, and each one of them has an implementation so we can support multiple engines in the future (currently only Ruby is fully supported). In this case you can have a look at how `org.asciidoctor.ast.Document` correlates to `org.asciidoctor.jruby.ast.impl.DocumentImp`.

All internal properties are already there during runtime, it's a matter of adding the methods to expose them.
RubyObjectWrapper (parent class of `DocumentImpl`) have methods to facilitate manipulation and conversion between Ruby and Java objects, so luckily you should not need to mess much with the jRuby library.

-----------------------------

Now, I attempted a POC and found an issue, but I suspect this may be a limitation of the current parser.
The problem is that the footnotes are only accessible after conversion. I created a project with a Preprocessor, TreeProcessor, PostProcessor and a Converter. And the footnotes are only available in the PostProcessor with the code below.

```java
  if (node instanceof Document) {
      Document document = (Document) node;
      Object references = ((DocumentImpl) document).getProperty("references");
      RubyHash rubyReferences = (RubyHash) references;
    . . . .
  } else if (node instanceof Section) {
```
![image](https://user-images.githubusercontent.com/5781153/95683404-4b17c080-0beb-11eb-8e79-8ca2a1755b89.png)

I saw in the code that the footnotes are added to the catalog during macros substitution :% If anyone has any idea, would be appreciated.

**Comment by lread** [comment]
**footnotes**
Hmm... my debugger shows me the same for `footnotes`.

I am frankly too green to understand what gets called from where, but I do see the `footnotes` seem to be populated on the Ruby side here: https://github.com/asciidoctor/asciidoctor/blob/5845c665c8f8eec0c95154b00adc32c383fc4efc/lib/asciidoctor/substitutors.rb#L851-L868

**api**
It is subjective, so take this with a grain of salt, but my preference would be to put `getFootnotes()` and `getRefs()` directly under the `document` because it keeps a symmetry between the Java and Ruby APIs.

**effort**
Also if you are happily into this already, I'm good with standing aside.  
I happy to help if there is a need.

**Comment by abelsromero** [comment]
> I do see the `footnotes` seem to be populated on the Ruby side here

I got there too, and if the footnotes are efetively processed after parsing, I suspect that even with the API changes they won't be available in the Converter :disappointed:

> It is subjective, so take this with a grain of salt, but my preference would be to put `getFootnotes()` and `getRefs()` directly under the `document` because it keeps a symmetry between the Java and Ruby APIs.

Then, it's up to @robertpanzer to break the tie. I really have no strong opinion.


> Also if you are happily into this already, I'm good with standing aside.
> I happy to help if there is a need.

Tbh, there's no hurry and I have other things. The first issue is always the hardest, just take your time.
You can try modifying the `a_postprocessor_instance_should_be_executed_after_document_is_rendered` test and associated PostProcessor to do something with the footnotes. If you have trouble and are still interested (no compromise at all), we could make an online pairing session.

**Comment by robertpanzer** [comment]
Good catch and we should definitely add it!

If you like to hack on this I am happy to help. 
Otherwise I can take an attempt too, I cannot give any ETA though.

**Comment by lread** [comment]
Thanks, I'll take a stab at it then.

**Comment by lread** [comment]
I am making some progress.

**footnotes**
So @abelsromero, the footnotes get populated by asciidoctor, but only as the adoc is converted.  
Which is fine, I guess, because footnotes are rendered at the end of a doc.  
So I think I have a handle on footnotes, although I'm sure you folks will have some feedback/tweaks on how I implemented.

**refs - how are they used?**
As to `refs`, I first need to understand how they are being used in the asciidoctor html5 converter (with the assumption that my, and other converters, will need to do this as well).
https://github.com/asciidoctor/asciidoctor/blob/cabde9c25fe040221ad71bb6aaf8754d2e5ec099/lib/asciidoctor/converter/html5.rb#L1115-L1124
```ruby
          if AbstractNode === (ref = (@refs ||= node.document.catalog[:refs])[refid]) && (@resolving_xref ||= (outer = true)) && outer
            if !(text = ref.xreftext node.attr 'xrefstyle', nil, true)
              text = %([#{refid}])
            elsif text.include? '<a'
              text = text.gsub DropAnchorRx, ''
            end
            @resolving_xref = nil
          else
            text = %([#{refid}])
          end
```

It seems this code got a bit more complicated to [address a circular reference problem](https://github.com/asciidoctor/asciidoctor/commit/05b8e7cca16976108927eb190e8f3471f6688421).

Putting such details in a converter seems a bit odd to me, it seems like all converters would want this work to be done on their behalf, but what do I know?  I certainly don't have a deep understanding yet here! 🙂 

Forgive my Ruby newbieness: So far I've figured out `===` means "instance of?", and that `=|` is a conditional assignment operator which means "if I am not already set (or falsey), then set me to", and that `@` is a class instance variable.  It looks like `@refs` is simply used for caching catalog refs.  `@resolving_xref` will be true if if ref is found, but `outer` will only be true if `@resolving_xref` was not already true.  

So I think this pseudocode might reflect what is effectively happening?

     if ref with refid found,
        and it is an asciidoctor node, 
        and we did not find a ref the very last time we checked
        and ref has reftext
     then
         text = reftext but with <a>.*</a> removed
     else
         text = [$refid]
     endif



**where should to model refs?**
Also of interest, is that `html5.rb` accesses `refs` from under the `catalog`.  Unlike `footnotes`, there is no `refs` duplicated directly off of `Document`.  So @abelsromero, maybe your idea of:

- `Document`
    - `getReferences()`
        - `getFootnotes()`
        - `getRefs()`

Is the way to go?  `getReferences().getRefs()` is a bit weird though.  Maybe `getCatalog()` instead of `getReferences()`?.

**Comment by abelsromero** [comment]
> Which is fine, I guess, because footnotes are rendered at the end of a doc.

Not related to the position, but this a known limitation of the current parser. Ideally the parser should provide the full AST with all elements right from the beginning. However, macros and includes (at least, that i know of) are parsed during conversion, so, any extension or converter that runs before that stage does not have that information. This is something to be addressed in the long future :%
We can only mention that in the integrator's guide, there's really not much we can do.

To manipulate the integration these can help:
* https://robertpanzer.github.io/blog/2015/jrubymethods.html
* https://robertpanzer.github.io/blog/2014/jrubydynamics.html

> Is the way to go? `getReferences().getRefs()` is a bit weird though. Maybe `getCatalog()` instead of `getReferences()`?.

I always assumed we have some space to divert from the core. The whole idea of Asciidoctorj was to abstract and make asciidoctor Java-idiomatic.
All these elements belong to the document semantically, and having them at that level make the API easy to use. If we all agree, and the naming does not create confusion with other elements, I think is fine. You correctly point that references.refs is weird, and also there's a concept that appears from time to time which is the "images catalog", so I am not in favor of catalog either.
In short, the proposal of putting them at document is fine.

I cannot help much on the other points you mention, I am not Ruby expert either. If we agree to place the methods at Document, you can focus on the footnotes for now, and when is done revisit the refs?

**Comment by lread** [comment]
Much thanks for the continued interest and replies @abelsromero!

> If we agree to place the methods at Document, ...

Yup, that works for me.

> ... you can focus on the footnotes for now, and when is done revisit the refs?

That's a great idea and will allow me to get earlier feedback on code.

**Comment by mojavelinux** [comment]
> Maybe getCatalog() instead of getReferences()?

I would say yes. It's best to keep with the terminology used by core here.

Also keep in mind that the catalog is very much an internal table intended for use in converters, but subject to change.

**Comment by lread** [comment]
Ah, ok thanks @mojavelinux (and @abelsromero for the 👍), so that implies we'll go with:

- `Document`
    - `getCatalog()`
        - `getFootnotes()`
        - `getRefs()`

> Also keep in mind that the catalog is very much an internal table intended for use in converters, but subject to change.

Fair enough, thanks for the heads up. Because the Ruby API also provides access to footnotes directly under Document, does that mean it is considered less internal and subject to change? If so maybe?:

- `Document`
    - `getFootnotes()`
    - `getCatalog()`
        - `getRefs()`

We could even mimic Ruby API more closely via:

- `Document`
    - `getFootnotes()`
    - `getCatalog()`
        - `getFootnotes()`
        - `getRefs()`

I am happy to comply with whatever makes sense to you folks.

**Comment by mojavelinux** [comment]
> Because the Ruby API also provides access to footnotes directly under Document, does that mean it is considered less internal and subject to change

The footnotes method is merely a convenience for the converters. It still looks in the references catalog and is thus part of said catalog. How you want to expose these references is up to the AsciidoctorJ project.

What I mean by them being subject to change is that the catalog is an extremely minimal catalog for the purpose of completing the conversion. They are not intended to be a general metamodel for the document. Thus, what they contain is biased towards the needs of the converter.

**Comment by lread** [comment]
Thanks @mojavelinux, that helps.

To AsciidoctorJ folks with an interest, I'll start with:

- `Document`
    - `getCatalog()`
        - `getFootnotes()`
        - `getRefs()`
